// @sap/bas-sdk/dist/test/apis/package-action.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const index_1 = require("../../src/index");
const axios_1 = __importDefault(require("axios"));
const axios_mock_adapter_1 = __importDefault(require("axios-mock-adapter"));
describe("package action scope", () => {
    let mockAxios;
    before(() => {
        mockAxios = new axios_mock_adapter_1.default(axios_1.default);
    });
    after(() => {
        mockAxios.restore();
    });
    describe("createActionRequest", () => {
        const startupServerUrl = "https://startup-test.domain.example.com/";
        const actionUrl = `${startupServerUrl}action`;
        afterEach(() => {
            mockAxios.reset();
        });
        it("createActionRequest success", async () => {
            const mockResponse = {
                success: true,
                message: "Action executed successfully",
            };
            mockAxios.onPost(actionUrl).reply((config) => {
                var _a, _b;
                // Verify the request was configured correctly
                (0, chai_1.expect)((_a = config.headers) === null || _a === void 0 ? void 0 : _a["Content-Type"]).to.equal("application/x-www-form-urlencoded");
                (0, chai_1.expect)((_b = config.headers) === null || _b === void 0 ? void 0 : _b["Authorization"]).to.equal("Bearer test-jwt-token");
                (0, chai_1.expect)(config.data).to.equal("key1=value1&key2=value2");
                return [200, mockResponse];
            });
            const res = await index_1.core.createActionRequest({
                startupServerUrl,
                jwt: "test-jwt-token",
                form: {
                    key1: "value1",
                    key2: "value2",
                },
                enforceSecurityVerification: true,
            });
            (0, chai_1.expect)(res.status).to.equal(200);
            (0, chai_1.expect)(res.data).to.deep.equal(mockResponse);
        });
        it("createActionRequest with URL without trailing slash", async () => {
            const urlWithoutSlash = "https://startup-test.domain.example.com";
            const mockResponse = { success: true };
            mockAxios.onPost(`${urlWithoutSlash}/action`).reply(200, mockResponse);
            const res = await index_1.core.createActionRequest({
                startupServerUrl: urlWithoutSlash,
                jwt: "test-jwt-token",
                form: { actionName: "install" },
                enforceSecurityVerification: false,
            });
            (0, chai_1.expect)(res.status).to.equal(200);
            (0, chai_1.expect)(res.data).to.deep.equal(mockResponse);
        });
        it("sendActionSetupRequest handles error response", async () => {
            var _a, _b;
            const errorResponse = {
                error: "Action failed",
                message: "Package not found",
            };
            mockAxios.onPost(actionUrl).reply(404, errorResponse);
            try {
                await index_1.core.createActionRequest({
                    startupServerUrl,
                    jwt: "test-jwt-token",
                    form: { actionName: "uninstall" },
                    enforceSecurityVerification: true,
                });
                chai_1.expect.fail("Should have thrown an error");
            }
            catch (error) {
                (0, chai_1.expect)(axios_1.default.isAxiosError(error)).to.be.true;
                if (axios_1.default.isAxiosError(error)) {
                    (0, chai_1.expect)((_a = error.response) === null || _a === void 0 ? void 0 : _a.status).to.equal(404);
                    (0, chai_1.expect)((_b = error.response) === null || _b === void 0 ? void 0 : _b.data).to.deep.equal(errorResponse);
                }
            }
        });
        it("createActionRequest with default enforceSecurityVerification", async () => {
            const mockResponse = { success: true };
            mockAxios.onPost(actionUrl).reply((config) => {
                // Verify security verification is enabled by default
                (0, chai_1.expect)(config.httpsAgent).to.exist;
                return [200, mockResponse];
            });
            const res = await index_1.core.createActionRequest({
                startupServerUrl,
                jwt: "test-jwt-token",
                form: { actionName: "update" },
            });
            (0, chai_1.expect)(res.status).to.equal(200);
            (0, chai_1.expect)(res.data).to.deep.equal(mockResponse);
        });
    });
});
