// @sap-ux/adp-tooling/dist/writer/cf.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateCf = generateCf;
exports.writeUi5AppInfo = writeUi5AppInfo;
exports.generateCfConfig = generateCfConfig;
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = require("node:path");
const mem_fs_1 = require("mem-fs");
const mem_fs_editor_1 = require("mem-fs-editor");
const project_access_1 = require("@sap-ux/project-access");
const cf_1 = require("../cf");
const source_1 = require("../source");
const manifest_1 = require("./manifest");
const project_utils_1 = require("./project-utils");
const i18n_1 = require("./i18n");
const helper_1 = require("../base/helper");
const project_builder_1 = require("../base/project-builder");
/**
 * Writes the CF adp-project template to the mem-fs-editor instance.
 *
 * @param {string} basePath - The base path.
 * @param {CfAdpWriterConfig} config - The CF writer configuration.
 * @param {ToolsLogger} logger - The logger.
 * @param {Editor} fs - The memfs editor instance.
 * @returns {Promise<Editor>} The updated memfs editor instance.
 */
async function generateCf(basePath, config, logger, fs) {
    if (!fs) {
        fs = (0, mem_fs_editor_1.create)((0, mem_fs_1.create)());
    }
    const fullConfig = setDefaults(config);
    const { app, cf, ui5, project } = fullConfig;
    await (0, cf_1.adjustMtaYaml)({
        projectPath: basePath,
        adpProjectName: project.name,
        appRouterType: cf.approuter,
        businessSolutionName: cf.businessSolutionName ?? '',
        businessService: cf.businessService,
        serviceKeys: cf.serviceInfo?.serviceKeys
    }, fs, config.options?.templatePathOverwrite, logger);
    if (fullConfig.app.i18nModels) {
        (0, i18n_1.writeI18nModels)(basePath, fullConfig.app.i18nModels, fs);
    }
    const variant = (0, project_utils_1.getCfVariant)(fullConfig);
    (0, manifest_1.fillDescriptorContent)(variant.content, app.appType, ui5.version, app.i18nModels);
    await (0, project_utils_1.writeCfTemplates)(basePath, variant, fullConfig, fs);
    await (0, project_utils_1.writeCfUI5Yaml)(fullConfig.project.folder, fullConfig, fs);
    return fs;
}
/**
 * Set default values for CF configuration.
 *
 * @param {CfAdpWriterConfig} config - The CF configuration provided by the calling middleware.
 * @returns {CfAdpWriterConfig} The enhanced configuration with default values.
 */
function setDefaults(config) {
    const configWithDefaults = {
        ...config,
        app: {
            ...config.app,
            appType: config.app.appType ?? (0, source_1.getApplicationType)(config.app.manifest),
            i18nModels: config.app.i18nModels ?? (0, i18n_1.getI18nModels)(config.app.manifest, config.app.layer, config.app.id),
            i18nDescription: config.app.i18nDescription ?? (0, i18n_1.getI18nDescription)(config.app.layer, config.app.title)
        },
        options: {
            addStandaloneApprouter: false,
            addSecurity: false,
            ...config.options
        }
    };
    return configWithDefaults;
}
/**
 * Fetch ui5AppInfo.json and write it to the project root.
 *
 * @param basePath - path to application root
 * @param ui5AppInfo - ui5AppInfo.json content
 * @param logger - logger instance
 */
async function writeUi5AppInfo(basePath, ui5AppInfo, logger) {
    try {
        const ui5AppInfoTargetPath = (0, node_path_1.join)(basePath, 'ui5AppInfo.json');
        node_fs_1.default.writeFileSync(ui5AppInfoTargetPath, JSON.stringify(ui5AppInfo, null, 2), 'utf-8');
        logger?.info(`Written ui5AppInfo.json to ${basePath}`);
    }
    catch (error) {
        logger?.error(`Failed to process ui5AppInfo.json: ${error.message}`);
        throw error;
    }
}
/**
 * Generate CF configuration for an adaptation project.
 *
 * @param basePath - path to project root
 * @param yamlPath - path to the project configuration file in YAML format
 * @param cfConfig - CF configuration
 * @param logger - logger instance
 * @param fs - mem-fs editor instance
 * @returns updated mem-fs editor instance
 */
async function generateCfConfig(basePath, yamlPath, cfConfig, logger, fs) {
    fs ??= (0, mem_fs_editor_1.create)((0, mem_fs_1.create)());
    const ui5Config = await (0, project_access_1.readUi5Yaml)(basePath, yamlPath);
    const bundlerTask = ui5Config.findCustomTask('app-variant-bundler-build');
    const serviceInstanceName = bundlerTask?.configuration?.serviceInstanceName;
    if (!serviceInstanceName) {
        throw new Error('No serviceInstanceName found in app-variant-bundler-build configuration');
    }
    const serviceInfo = await (0, cf_1.getServiceInstanceKeys)({
        names: [serviceInstanceName]
    }, logger);
    if (!serviceInfo || serviceInfo.serviceKeys.length === 0) {
        throw new Error(`No service keys found for service instance: ${serviceInstanceName}`);
    }
    const appId = await (0, helper_1.getBaseAppId)(basePath);
    const appHostIds = (0, cf_1.getAppHostIds)(serviceInfo.serviceKeys);
    const ui5AppInfo = await (0, cf_1.getCfUi5AppInfo)(appId, appHostIds, cfConfig, logger);
    if (appHostIds.length === 0) {
        throw new Error('No app host IDs found in service keys.');
    }
    await writeUi5AppInfo(basePath, ui5AppInfo, logger);
    await (0, cf_1.addServeStaticMiddleware)(basePath, ui5Config, logger);
    await (0, project_builder_1.runBuild)(basePath, { ADP_BUILDER_MODE: 'preview' });
    (0, cf_1.addBackendProxyMiddleware)(basePath, ui5Config, serviceInfo.serviceKeys, logger);
    fs.write((0, node_path_1.join)(basePath, yamlPath), ui5Config.toString());
    return fs;
}
