// @sap-ux/adp-tooling/dist/writer/writer-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfig = getConfig;
exports.getUi5Config = getUi5Config;
exports.getCfConfig = getCfConfig;
const node_path_1 = require("node:path");
const axios_extension_1 = require("@sap-ux/axios-extension");
const ui5_1 = require("../ui5");
const abap_1 = require("../abap");
const types_1 = require("../types");
const i18n_1 = require("../i18n");
const source_1 = require("../source");
/**
 * Generates the configuration object for the Adaptation Project.
 *
 * @param {ConfigOptions} options - The configuration options.
 * @param {AbapServiceProvider} options.provider - The ABAP service provider instance.
 * @param {ConfigAnswers} options.configAnswers - User-provided configuration details (system, application, etc.).
 * @param {FlexLayer} options.layer - The FlexLayer indicating the deployment layer.
 * @param {object} options.defaults - Default project parameters.
 * @param {string} options.defaults.namespace - The default namespace to be used.
 * @param {Package} options.packageJson - The package.json information for generating custom configuration.
 * @param {ToolsLogger} options.logger - The logger for debugging and error logging.
 * @returns {Promise<AdpWriterConfig>} A promise that resolves to the generated ADP writer configuration.
 */
async function getConfig(options) {
    const { configAnswers, attributeAnswers, layer, logger, packageJson, provider, publicVersions, systemVersion, manifest, projectType, toolsId, keyUserChanges } = options;
    const ato = await provider.getAtoInfo();
    const operationsType = ato.operationsType ?? 'P';
    const target = await (0, abap_1.getProviderConfig)(configAnswers.system, logger);
    const isCloudSystem = await provider.isAbapCloud();
    const isCustomerBase = layer === "CUSTOMER_BASE" /* FlexLayer.CUSTOMER_BASE */;
    const ui5Version = isCloudSystem
        ? (0, ui5_1.getLatestVersion)(publicVersions)
        : (0, ui5_1.getVersionToBeUsed)(attributeAnswers.ui5Version, isCustomerBase, publicVersions);
    const { namespace, title, enableTypeScript } = attributeAnswers;
    const { application: { id, bspName }, fioriId, ach } = configAnswers;
    const app = {
        id: namespace,
        reference: id,
        layer,
        title,
        manifest,
        ach,
        fioriId
    };
    const supportedProject = await (0, source_1.getSupportedProject)(provider);
    if (projectType === axios_extension_1.AdaptationProjectType.CLOUD_READY) {
        const lrep = provider.getLayeredRepository();
        const { activeLanguages: languages } = await lrep.getSystemInfo();
        Object.assign(app, {
            bspName,
            languages
        });
    }
    const ui5 = getUi5Config(ui5Version, publicVersions, systemVersion);
    return {
        app,
        ui5,
        customConfig: {
            adp: {
                environment: operationsType,
                support: {
                    id: packageJson.name ?? '',
                    version: packageJson.version ?? '',
                    toolsId
                },
                projectType,
                supportedProject
            }
        },
        target,
        options: {
            fioriTools: true,
            enableTypeScript
        },
        keyUserChanges
    };
}
/**
 * Generates the configuration details required for a SAPUI5 application based on system and selected UI5 versions.
 *
 * @param {string} ui5Version - The selected UI5 version.
 * @param {UI5Version} publicVersions - The publicly available UI5 versions.
 * @param {string | undefined} systemVersion - The SAPUI5 version detected on the target system.
 * @returns {AdpWriterConfig['ui5']} An object containing the required UI5 configuration for the writer config.
 */
function getUi5Config(ui5Version, publicVersions, systemVersion) {
    return {
        minVersion: (0, ui5_1.getMinUI5VersionForManifest)(publicVersions, systemVersion),
        version: (0, ui5_1.getFormattedVersion)(ui5Version),
        frameworkUrl: (0, ui5_1.getOfficialBaseUI5VersionUrl)(ui5Version),
        shouldSetMinVersion: (0, ui5_1.shouldSetMinUI5Version)(systemVersion)
    };
}
/**
 * Create CF configuration from batch objects.
 *
 * @param {CreateCfConfigParams} params - The configuration parameters containing batch objects.
 * @returns {CfAdpWriterConfig} The CF configuration.
 */
function getCfConfig(params) {
    const baseApp = params.cfServicesAnswers.baseApp;
    if (!baseApp) {
        throw new Error((0, i18n_1.t)('errors.baseAppRequired'));
    }
    const ui5Version = (0, ui5_1.getLatestVersion)(params.publicVersions);
    return {
        app: {
            id: baseApp.appId,
            title: params.attributeAnswers.title,
            layer: params.layer,
            namespace: params.attributeAnswers.namespace,
            manifest: params.manifest
        },
        baseApp,
        cf: {
            url: params.cfConfig.url,
            org: params.cfConfig.org,
            space: params.cfConfig.space,
            html5RepoRuntimeGuid: params.html5RepoRuntimeGuid,
            approuter: params.cfServicesAnswers.approuter ?? types_1.AppRouterType.MANAGED,
            businessService: params.cfServicesAnswers.businessService ?? '',
            businessSolutionName: params.cfServicesAnswers.businessSolutionName,
            serviceInstanceGuid: params.serviceInstanceGuid,
            backendUrls: params.backendUrls,
            oauthPaths: params.oauthPaths,
            serviceInfo: params.serviceInfo
        },
        project: {
            name: params.attributeAnswers.projectName,
            path: params.projectPath,
            folder: (0, node_path_1.join)(params.projectPath, params.attributeAnswers.projectName)
        },
        customConfig: {
            adp: {
                support: {
                    id: params.packageJson?.name ?? '',
                    version: params.packageJson?.version ?? '',
                    toolsId: params.toolsId ?? ''
                }
            }
        },
        ui5: {
            version: ui5Version
        },
        options: {
            addStandaloneApprouter: params.cfServicesAnswers.approuter === types_1.AppRouterType.STANDALONE
        }
    };
}
