// @sap-ux/adp-tooling/dist/cf/app/discovery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAppHostIds = getAppHostIds;
exports.getBackendUrlsFromServiceKeys = getBackendUrlsFromServiceKeys;
exports.getBackendUrlsWithPaths = getBackendUrlsWithPaths;
exports.getServiceKeyDestinations = getServiceKeyDestinations;
exports.getOAuthPathsFromXsApp = getOAuthPathsFromXsApp;
exports.getCfApps = getCfApps;
const node_fs_1 = require("node:fs");
const node_path_1 = require("node:path");
const i18n_1 = require("../../i18n");
const utils_1 = require("../utils");
const api_1 = require("../services/api");
/**
 * Get the app host ids from service keys.
 *
 * @param {ServiceKeys[]} serviceKeys - The service keys.
 * @returns {string[]} The app host ids.
 */
function getAppHostIds(serviceKeys) {
    const appHostIds = [];
    for (const serviceKey of serviceKeys) {
        const appHostId = serviceKey.credentials['html5-apps-repo']?.app_host_id;
        if (appHostId) {
            // There might be multiple appHostIds separated by comma
            const ids = appHostId.split(',').map((item) => item.trim());
            appHostIds.push(...ids);
        }
    }
    return [...new Set(appHostIds)];
}
/**
 * Extracts all backend URLs from service key credentials. Iterates through all endpoint keys to find all endpoints with URLs.
 * Handles both string endpoints and object endpoints with url property.
 *
 * @param {ServiceKeys[]} serviceKeys - The credentials from service keys.
 * @returns {string[]} Array of backend URLs (including full paths) or empty array if none found.
 */
function getBackendUrlsFromServiceKeys(serviceKeys) {
    if (!serviceKeys || serviceKeys.length === 0) {
        return [];
    }
    const urls = [];
    const endpoints = serviceKeys[0]?.credentials?.endpoints;
    if (endpoints && typeof endpoints === 'object' && endpoints !== null) {
        for (const key in endpoints) {
            const endpoint = endpoints[key];
            if (endpoint?.url) {
                urls.push(endpoint.url);
            }
        }
    }
    return urls;
}
/**
 * Extract destination to URL mapping from service key endpoints.
 *
 * @param {ServiceKeys[]} serviceKeys - The service keys containing endpoints.
 * @returns {Map<string, string>} Map of destination names to URLs.
 */
function extractDestinationToUrlMap(serviceKeys) {
    const destinationToUrl = new Map();
    const endpoints = serviceKeys[0]?.credentials?.endpoints;
    if (endpoints && typeof endpoints === 'object') {
        for (const key in endpoints) {
            const endpoint = endpoints[key];
            if (endpoint?.url && endpoint.destination) {
                destinationToUrl.set(endpoint.destination, endpoint.url);
            }
        }
    }
    return destinationToUrl;
}
/**
 * Clean regex pattern from route source.
 *
 * @param {string} source - The route source pattern.
 * @returns {string} Cleaned path.
 */
function cleanRoutePath(source) {
    let path = source;
    // Remove leading ^ and trailing $
    path = path.replace(/^\^/, '').replace(/\$$/, '');
    // Remove capture groups like (.*) or $1
    path = path.replace(/\([^)]*\)/g, '');
    // Remove regex quantifiers
    path = path.replace(/\$\d+/g, '');
    // Clean up any remaining regex characters at the end
    path = path.replace(/\/?\*$/, '');
    // Normalize multiple consecutive slashes to single slash
    while (path.includes('//')) {
        path = path.replaceAll('//', '/');
    }
    return path;
}
/**
 * Process a route and add its cleaned path to the destination mapping.
 *
 * @param {XsApp['routes'][number]} route - The route object from xs-app.json.
 * @param {Map<string, Set<string>>} destinationToPaths - Map to store destination to paths mapping.
 */
function processRouteForDestination(route, destinationToPaths) {
    const destination = route.destination;
    const service = route.service;
    if (!destination || service === 'html5-apps-repo-rt' || !route.source) {
        return;
    }
    const path = cleanRoutePath(route.source);
    if (path) {
        if (!destinationToPaths.has(destination)) {
            destinationToPaths.set(destination, new Set());
        }
        destinationToPaths.get(destination).add(path);
    }
}
/**
 * Extract destination to paths mapping from xs-app.json routes.
 *
 * @param {string} xsAppPath - Path to xs-app.json file.
 * @returns {Map<string, Set<string>>} Map of destination names to path sets.
 */
function extractDestinationToPathsMap(xsAppPath) {
    const destinationToPaths = new Map();
    try {
        const xsAppContent = (0, node_fs_1.readFileSync)(xsAppPath, 'utf8');
        const xsApp = JSON.parse(xsAppContent);
        if (xsApp?.routes) {
            for (const route of xsApp.routes) {
                processRouteForDestination(route, destinationToPaths);
            }
        }
    }
    catch (e) {
        throw new Error((0, i18n_1.t)('error.invalidXsAppJson', { error: e.message }));
    }
    return destinationToPaths;
}
/**
 * Maps backend URLs to their corresponding OAuth paths based on destination matching
 * between xs-app.json routes and credentials.json endpoints.
 *
 * @param {ServiceKeys[]} serviceKeys - The service keys containing endpoints with destinations.
 * @param {string} basePath - Path to the .adp/reuse folder containing xs-app.json files.
 * @returns {Array<{ url: string; paths: string[] }>} Array of URL-to-paths mappings.
 */
function getBackendUrlsWithPaths(serviceKeys, basePath) {
    const destinationToUrl = extractDestinationToUrlMap(serviceKeys);
    const reuseXsAppPath = (0, node_path_1.join)(basePath, '.adp', 'reuse', 'xs-app.json');
    const distXsAppPath = (0, node_path_1.join)(basePath, 'dist', 'xs-app.json');
    let xsAppPath;
    if ((0, node_fs_1.existsSync)(reuseXsAppPath)) {
        xsAppPath = reuseXsAppPath;
    }
    else if ((0, node_fs_1.existsSync)(distXsAppPath)) {
        xsAppPath = distXsAppPath;
    }
    else {
        throw new Error((0, i18n_1.t)('error.xsAppJsonNotFound', { paths: `${reuseXsAppPath}, ${distXsAppPath}` }));
    }
    const destinationToPaths = extractDestinationToPathsMap(xsAppPath);
    const result = [];
    for (const [destination, paths] of destinationToPaths.entries()) {
        const url = destinationToUrl.get(destination);
        if (url) {
            result.push({
                url,
                paths: Array.from(paths)
            });
        }
    }
    return result;
}
/**
 * Extract endpoint destinations from service keys.
 *
 * @param {ServiceKeys[]} serviceKeys - The service keys.
 * @returns {Array<{name: string; url: string}>} Array of endpoint destinations.
 */
function getServiceKeyDestinations(serviceKeys) {
    const endpointDestinations = [];
    for (const key of serviceKeys) {
        const endpoints = key.credentials?.endpoints;
        if (endpoints && typeof endpoints === 'object') {
            for (const endpointKey in endpoints) {
                const endpoint = endpoints[endpointKey];
                if (endpoint?.url && endpoint.destination) {
                    endpointDestinations.push({
                        name: endpoint.destination,
                        url: endpoint.url
                    });
                }
            }
        }
    }
    return endpointDestinations;
}
/**
 * Extracts OAuth paths from xs-app.json routes that have a source property.
 * These paths should receive OAuth Bearer tokens in the middleware.
 *
 * @param {AdmZip.IZipEntry[]} zipEntries - The zip entries containing xs-app.json.
 * @returns {string[]} Array of path patterns (from route.source) that have a source property.
 */
function getOAuthPathsFromXsApp(zipEntries) {
    const xsApp = (0, utils_1.extractXSApp)(zipEntries);
    if (!xsApp?.routes) {
        return [];
    }
    const pathsSet = new Set();
    for (const route of xsApp.routes) {
        if (route.service === 'html5-apps-repo-rt' || !route.source) {
            continue;
        }
        const path = cleanRoutePath(route.source);
        if (path) {
            pathsSet.add(path);
        }
    }
    return Array.from(pathsSet);
}
/**
 * Discover apps from FDC API based on credentials.
 *
 * @param {ServiceKeys[]} serviceKeys - The service keys containing app host IDs
 * @param {CfConfig} cfConfig - The CF configuration
 * @param {ToolsLogger} logger - The logger
 * @returns {Promise<CFApp[]>} The discovered apps
 */
async function getCfApps(serviceKeys, cfConfig, logger) {
    const appHostIds = getAppHostIds(serviceKeys);
    logger?.log(`App Host Ids: ${JSON.stringify(appHostIds)}`);
    // Validate appHostIds array length (max 100 as per API specification)
    if (appHostIds.length > 100) {
        throw new Error((0, i18n_1.t)('error.tooManyAppHostIds', { appHostIdsLength: appHostIds.length }));
    }
    return (0, api_1.getFDCApps)(appHostIds, cfConfig, logger);
}
