// @sap-ux/adp-tooling/dist/base/credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.storeCredentials = storeCredentials;
const store_1 = require("@sap-ux/store");
/**
 * Stores system credentials securely using the @sap-ux/store service.
 * Only stores credentials for ABAP environments when all required fields are provided.
 *
 * @param {ConfigAnswers} configAnswers - Configuration answers containing credentials and system info
 * @param {SystemLookup} systemLookup - System lookup service for retrieving endpoint details
 * @param {ToolsLogger} logger - Logger for informational and warning messages
 * @returns {Promise<void>} Promise that resolves when credentials are stored or operation completes
 */
async function storeCredentials(configAnswers, systemLookup, logger) {
    if (!configAnswers.username || !configAnswers.password) {
        return;
    }
    try {
        const systemEndpoint = await systemLookup.getSystemByName(configAnswers.system);
        if (!systemEndpoint?.Url) {
            logger.warn('Cannot store credentials: system endpoint or URL not found.');
            return;
        }
        const systemService = await (0, store_1.getService)({
            entityName: 'system'
        });
        const backendSystemKey = new store_1.BackendSystemKey({
            url: systemEndpoint.Url,
            client: systemEndpoint.Client
        });
        const existingSystem = await systemService.read(backendSystemKey);
        const backendSystem = new store_1.BackendSystem({
            name: configAnswers.system,
            url: systemEndpoint.Url,
            client: systemEndpoint.Client,
            username: configAnswers.username,
            password: configAnswers.password,
            systemType: systemEndpoint.SystemType || store_1.SystemType.AbapOnPrem,
            connectionType: 'abap_catalog',
            userDisplayName: configAnswers.username
        });
        await systemService.write(backendSystem, { force: !!existingSystem });
        logger.info('System credentials have been stored securely.');
    }
    catch (error) {
        logger.error(`Failed to store credentials: ${error instanceof Error ? error.message : String(error)}`);
        logger.debug(error);
    }
}
